File: benchmarks/lib/manifest.mjs
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";

/** The only metrics a case may declare. */
export const METRICS = ["cold-start", "first-run", "warm", "streaming"];

/**
 * Load and validate corpus/manifest.json.
 * @param {string} corpusDir absolute path to the corpus directory
 */
export function loadManifest(corpusDir) {
  const raw = JSON.parse(readFileSync(join(corpusDir, "manifest.json"), "utf-8"));
  if (!Array.isArray(raw.cases)) throw new Error("manifest.cases must be an array");
  const ids = new Set();
  for (const c of raw.cases) {
    if (!c.id) throw new Error("manifest case is missing an id");
    if (ids.has(c.id)) throw new Error(`duplicate case id: ${c.id}`);
    ids.add(c.id);
    if (!Array.isArray(c.metrics) || c.metrics.length === 0) {
      throw new Error(`case ${c.id} must declare a non-empty metrics[]`);
    }
    for (const metric of c.metrics) {
      if (!METRICS.includes(metric)) throw new Error(`case ${c.id} has unknown metric: ${metric}`);
    }
    if (!c.script || !existsSync(join(corpusDir, c.script))) {
      throw new Error(`case ${c.id} script not found: ${c.script}`);
    }
    for (const [name, inp] of Object.entries(c.inputs ?? {})) {
      if (inp.file && !inp.generated && !existsSync(join(corpusDir, inp.file))) {
        throw new Error(`case ${c.id} input '${name}' file not found: ${inp.file}`);
      }
    }
  }
  return { corpusDir, cases: raw.cases, ids };
}

/** Cases whose declared metrics[] includes `metric`. */
export function casesForMetric(manifest, metric) {
  return manifest.cases.filter((c) => c.metrics.includes(metric));
}

/**
 * Read a case's declared inputs into Buffers.
 * @returns {Record<string,{buffer:Buffer, mimeType:string, charset?:string}>}
 */
export function resolveInputs(manifest, caseObj) {
  const out = {};
  for (const [name, inp] of Object.entries(caseObj.inputs ?? {})) {
    const buffer = readFileSync(join(manifest.corpusDir, inp.file));
    out[name] = { buffer, mimeType: inp.mimeType, charset: inp.charset };
  }
  return out;
}

/** Fail-fast: throw if any result case carries an id not present in the manifest. */
export function validateResultIds(manifest, resultCases) {
  for (const rc of resultCases) {
    if (!manifest.ids.has(rc.id)) {
      throw new Error(`result contains orphan id not in manifest: ${rc.id}`);
    }
  }
}

File: benchmarks/runners/python/manifest.py
"""Parse and validate corpus/manifest.json, resolve scripts + inputs.
Mirrors benchmarks/lib/manifest.mjs."""

import json
from pathlib import Path

# The only metrics a case may declare.
METRICS = ("cold-start", "first-run", "warm", "streaming")


def load_manifest(corpus_dir):
    corpus_dir = Path(corpus_dir)
    raw = json.loads((corpus_dir / "manifest.json").read_text(encoding="utf-8"))
    cases = raw.get("cases")
    if not isinstance(cases, list):
        raise ValueError("manifest.cases must be an array")
    ids = set()
    for c in cases:
        cid = c.get("id")
        if not cid:
            raise ValueError("manifest case is missing an id")
        if cid in ids:
            raise ValueError(f"duplicate case id: {cid}")
        ids.add(cid)
        metrics = c.get("metrics")
        if not isinstance(metrics, list) or not metrics:
            raise ValueError(f"case {cid} must declare a non-empty metrics[]")
        for m in metrics:
            if m not in METRICS:
                raise ValueError(f"case {cid} has unknown metric: {m}")
        script = c.get("script")
        if not script or not (corpus_dir / script).exists():
            raise ValueError(f"case {cid} script not found: {script}")
        for name, inp in (c.get("inputs") or {}).items():
            if inp.get("file") and not inp.get("generated") \
                    and not (corpus_dir / inp["file"]).exists():
                raise ValueError(f"case {cid} input '{name}' file not found: {inp['file']}")
    return {"corpusDir": str(corpus_dir), "cases": cases, "ids": ids}


def cases_for_metric(manifest, metric):
    return [c for c in manifest["cases"] if metric in c["metrics"]]


def resolve_inputs(manifest, case_obj):
    """Read a case's declared inputs into bytes."""
    corpus_dir = Path(manifest["corpusDir"])
    out = {}
    for name, inp in (case_obj.get("inputs") or {}).items():
        data = (corpus_dir / inp["file"]).read_bytes()
        out[name] = {"bytes": data, "mimeType": inp["mimeType"], "charset": inp.get("charset")}
    return out


def read_script(manifest, case_obj):
    return (Path(manifest["corpusDir"]) / case_obj["script"]).read_text(encoding="utf-8")


def validate_result_ids(manifest, result_ids):
    """Fail-fast: raise if any result id is not present in the manifest."""
    for rid in result_ids:
        if rid not in manifest["ids"]:
            raise ValueError(f"result contains orphan id not in manifest: {rid}")
